elastic/kibana
# [Fleet] Simplified package policy create API, enriching default values

## Summary

Closes https://github.com/elastic/kibana/issues/113750

Made most properties optional in `POST /api/fleet/package_policies`, loading default values from base package.
This is not a breaking change.

Remaining tasks:
- add/update unit tests - DONE
- update openapi spec - DONE

To verify:

- send minimal request to API to check package policy is created with correct defaults
- if one input is set to `enabled: false`, it is propagated down to all `inputs.streams`
- if `package.version` does not match the installed base package version, there is an error coming
e.g. 
```
POST http://elastic:changeme@localhost:5601/julia/api/fleet/package_policies
kbn-xsrf: kibana

{"name":"apache-5","inputs":[{"type":"logfile","enabled":true},{"type":"apache/metrics","enabled":false}],"package":{"name":"apache","version":"0.3.3"}}
```

![image](https://user-images.githubusercontent.com/90178898/143463048-32b0f4d0-5f4d-4d6c-8969-e5090f4dc847.png)

- create package policy from UI and customize values (agent_policy, input stream values, etc.) and verify custom values are persisted
![image](https://user-images.githubusercontent.com/90178898/143464179-4b23dbc6-0476-4087-9caa-b40000463f8f.png)

Update on Nov 30, simplified update API. To verify:
- create a package policy, note down the policy id
- send a PUT request with minimal body (no required fields here) e.g. update namespace or description
- verify that request was successful and only modified field has changed
- the API allows resetting values as well e.g. `description: "", enabled: false`
- example:
```
PUT http://elastic:changeme@localhost:5620/api/fleet/package_policies/64f13192-5dca-42f3-aa9c-62bc94c2f12c
kbn-xsrf: kibana

{ "description":"desc"}
```

Verify updated open api spec here:
- https://petstore.swagger.io/?url=https://raw.githubusercontent.com/elastic/kibana/72f8fb29f1746e8e2b4917793df387119bf8757f/x-pack/plugins/fleet/common/openapi/bundled.json#/default/create-package-policy
- https://petstore.swagger.io/?url=https://raw.githubusercontent.com/elastic/kibana/72f8fb29f1746e8e2b4917793df387119bf8757f/x-pack/plugins/fleet/common/openapi/bundled.json#/default/update-package-policy

### Checklist


### For maintainers
